src/nbl/video/IGPUDescriptorSet.cpp

```diff
@@ -5,81 +5,64 @@
-        break;
-    case asset::EDT_UNIFORM_BUFFER_DYNAMIC:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + (m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER]);
-        break;
-    case asset::EDT_STORAGE_BUFFER_DYNAMIC:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + (m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC]);
-        break;
-    case asset::EDT_INPUT_ATTACHMENT:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_storageImageStorage.get()) + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE];
-        break;
-    case asset::EDT_ACCELERATION_STRUCTURE:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_accelerationStructureStorage.get());
-        break;
-    default:
-        assert(!"Invalid code path.");
-        return nullptr;
-    }
+        const bool allowsFreeing = m_pool->m_creationParameters.flags.hasFlags(IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT);
+        for (auto i = 0u; i < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++i)
+        {
+            // There is no descriptor of such type in the set.
+            if (m_descriptorStorageOffsets.data[i] == ~0u)
+                continue;
 
-    return baseAddress;
-}
+            const auto type = static_cast<asset::IDescriptor::E_TYPE>(i);
 
-core::smart_refctd_ptr<IGPUSampler>* IGPUDescriptorSet::getMutableSamplerStorage() const
-{
-    return reinterpret_cast<core::smart_refctd_ptr<IGPUSampler>*>(m_pool->m_mutableSamplerStorage.get());
+            const uint32_t allocatedOffset = getDescriptorStorageOffset(type);
+            assert(allocatedOffset != ~0u);
+
+            const uint32_t count = m_layout->getTotalDescriptorCount(type);
+            assert(count != 0u);
+
+            std::destroy_n(m_pool->getDescriptorStorage(type) + allocatedOffset, count);
+
+            if (allowsFreeing)
+                m_pool->m_descriptorAllocators[i]->free(allocatedOffset, count);
+        }
+
+        const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
+        if (mutableSamplerCount > 0)
+        {
+            const uint32_t allocatedOffset = getMutableSamplerStorageOffset();
+            assert(allocatedOffset != ~0u);
+
+            std::destroy_n(m_pool->getMutableSamplerStorage() + allocatedOffset, mutableSamplerCount);
+
+            if (allowsFreeing)
+                m_pool->m_descriptorAllocators[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)]->free(allocatedOffset, mutableSamplerCount);
+        }
+    }
```

Comment:
erm, where's the call to `vkFreeDescriptorSets` ?

Reply:
as in: what would call it?

Reply:
It should be called in `~CVulkanDescriptorSet` if the pool has the flag set which allows the call to `vkFreeDescriptorSets`.

Reply:
Resolved.